prediction_market_agent/agents/goal_manager.py

```diff
@@ -0,0 +1,295 @@
+        )
+
+    def to_goal(self) -> Goal:
+        return Goal(
+            goal=self.goal,
+            motivation=self.motivation,
+            completion_criteria=self.completion_criteria,
+        )
+
+
+class GoalManager:
+    def __init__(
+        self,
+        agent_id: str,
+        high_level_description: str,
+        agent_capabilities: str,
+        retry_limit: int = 3,
+        model: str = DEFAULT_OPENAI_MODEL,
+        sqlalchemy_db_url: str | None = None,
+    ):
+        self.agent_id = agent_id
+        self.high_level_description = high_level_description
+        self.agent_capabilities = agent_capabilities
+        self.retry_limit = retry_limit
+        self.model = model
+        self.table_handler = EvaluatedGoalTableHandler(
+            agent_id=agent_id,
+            sqlalchemy_db_url=sqlalchemy_db_url,
+        )
+
+    def get_latest_evaluated_goals_from_memory(self, limit: int) -> list[EvaluatedGoal]:
+        evaluated_goal_models = self.table_handler.get_latest_evaluated_goals(
+            limit=limit
+        )
+        return [EvaluatedGoal.from_model(model) for model in evaluated_goal_models]
+
+    def generate_goal(self, latest_evaluated_goals: list[EvaluatedGoal]) -> Goal:
+        """
+        Generate a new goal based on the high-level description and the latest
+        evaluated goals.
+
+        TODO support generation of long-horizon goals with a specified
+        completion date, until which the goal's status is 'pending'.
+        """
+        parser = PydanticOutputParser(pydantic_object=Goal)
+        prompt = PromptTemplate(
+            template=GENERATE_GOAL_PROMPT_TEMPLATE,
+            input_variables=[
+                "high_level_description",
+                "agent_capabilities",
+                "previous_evaluated_goals",
+            ],
+            partial_variables={"format_instructions": parser.get_format_instructions()},
+        )
+        latest_evaluated_goals_str = self.evaluated_goals_to_str(latest_evaluated_goals)
+        llm = ChatOpenAI(
+            temperature=0,
+            model=self.model,
+            api_key=APIKeys().openai_api_key_secretstr_v1,
+        )
+        chain = prompt | llm | parser
+
+        goal: Goal = chain.invoke(
+            {
+                "high_level_description": self.high_level_description,
+                "agent_capabilities": self.agent_capabilities,
+                "previous_evaluated_goals": latest_evaluated_goals_str,
+            }
+        )
+        return goal
+
+    def have_reached_retry_limit(
+        self, latest_evaluated_goals: list[EvaluatedGoal]
+    ) -> bool:
+        if self.retry_limit == 0:
+            return True
+
+        if len(latest_evaluated_goals) < self.retry_limit + 1:
+            return False
+
+        latest_goal = latest_evaluated_goals[0].to_goal()
+        return all(
+            g.to_goal() == latest_goal
+            for g in latest_evaluated_goals[: self.retry_limit + 1]
+        )
```

Comment:
Could we use tenacity without decorators (see [link](https://github.com/jd/tenacity/issues/155)) here?

Reply:
What's wrong with the decorator?

Reply:
I'm not sure how tenacity.retry would fit into this. The goal manager here is fetching previous goals from the db, and checking if the latest goal is duplicated `self.retry_limit` times. i.e. there's no try-except loop to replace.

Reply:
Understood that retry cannot be bundled with tenacity. Closing this.